android/src/main/java/com/reactnativeleveldown/LeveldownModule.java

```diff
@@ -0,0 +1,288 @@
+import com.github.hf.leveldb.util.SimpleWriteBatch;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+public class LeveldownModule extends ReactContextBaseJavaModule implements LifecycleEventListener {
+
+    private final ReactApplicationContext reactContext;
+
+    public LeveldownModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+        this.reactContext = reactContext;
+        reactContext.addLifecycleEventListener(this);
+    }
+
+    @Override
+    public String getName() {
+        return "Leveldown";
+    }
+
+    private static final String TAG = "ReactNativeLeveldown";
+
+    private static final String E_ALREADY_OPEN = "E_ALREADY_OPEN";
+    private static final String E_UNKNOWN_HANDLE = "E_UNKNOWN_HANDLE";
+    private static final String E_OPEN_ERROR = "E_OPEN_ERROR";
+    private static final String E_PUT_ERROR = "E_PUT_ERROR";
+    private static final String E_DELETE_ERROR = "E_DELETE_ERROR";
+    private static final String E_BATCH_UNKNOWN_OPERATION_TYPE = "E_BATCH_UNKNOWN_OPERATION_TYPE";
+    private static final String E_BATCH_OPERATION_ERROR = "E_BATCH_OPERATION_ERROR";
+    private static final String E_GET_ERROR = "E_GET_ERROR";
+    private static final String E_CLEAR_ERROR = "E_CLEAR_ERROR";
+    private static final String E_ITERATOR_CREATE = "E_ITERATOR_CREATE";
+    private static final String E_ALREADY_INITIALIZED = "E_ALREADY_INITIALIZED";
+    private static final String E_ITERATOR_GET = "E_ITERATOR_GET";
+    private static final String E_ITERATOR_SEEK = "E_ITERATOR_SEEK";
+    private static final String E_ITERATOR_CLOSE = "E_ITERATOR_SEEK";
+
+    private Map<Integer, LevelDB> dbHandleTable = new HashMap<>();
+
+    private LevelDB getDb(int dbHandle, Promise promise) {
+        if (!dbHandleTable.containsKey(dbHandle)) {
+            promise.reject(E_UNKNOWN_HANDLE,  String.format("Unknown DB handle %s", dbHandle));
+        }
+        return dbHandleTable.get(dbHandle);
+    }
+    private Map<Integer, LeveldownIterator> iteratorWrapperTable = new HashMap<>();
+
+    private LeveldownIterator getIterator(int iteratorHandle, Promise promise) {
+        if (!iteratorWrapperTable.containsKey(iteratorHandle)) {
+            promise.reject(E_UNKNOWN_HANDLE,  String.format("Unknown iterator handle %s", iteratorHandle));
+        }
+        return iteratorWrapperTable.get(iteratorHandle);
+    }
+
+    @ReactMethod
+    public void open(int dbHandle, String databaseName, boolean createIfMissing, boolean errorIfExists, Promise promise) {
+        if (dbHandleTable.containsKey(dbHandle)) {
+            promise.reject(E_ALREADY_OPEN,  String.format("DB with handle %s already open", dbHandle));
+        }
+
+        LevelDB.Configuration configuration = LevelDB.configure().createIfMissing(createIfMissing).exceptionIfExists(errorIfExists);
+
+        try {
+            LevelDB db = LevelDB.open(reactContext.getFilesDir().getAbsolutePath() +
```

Comment:
Since these databases aren't user-modifiable, would it make more sense to use `getDataDir`? My understanding is that files in this path would show up elsewhere in the Android interface—e.g. a file browser app. But I don't know what's idiomatic platform-wise here!

Reply:
I admit that I wasn't thoughtful about that, I just looked it up.
`getFilesDir` is the way to go.

`getFilesDir` is meant for app-specific data, which shouldn't be accessed from other apps. Sources: [Data and file storage overview](https://developer.android.com/training/data-storage) and [Access app-specific files](https://developer.android.com/training/data-storage/app-specific#java).

The [API](https://developer.android.com/reference/android/content/Context#getDataDir()) says that `getDataDir` should not be used, actually `getFilesDir` resides inside `getDataDir`, see [this stackoverflow question](https://stackoverflow.com/q/39920986/1046061).